hydra-node/src/Hydra/Network/Ouroboros.hs

```diff
@@ -129,34 +130,62 @@ withOuroborosNetwork tracer localHost remoteHosts networkCallback between = do
   let newBroadcastChannel = atomically $ dupTChan bchan
   -- TODO: Factor this out, there should be only one IOManager per process.
   withIOManager $ \iomgr -> do
-    race_ (connect iomgr newBroadcastChannel hydraClient) $
+    -- init a signal mvar used by Network.setPeers to communicate with thread 'connect'
+    signal <- newMVar remoteHosts
+    -- init read signal to avoid blocking on Network.getPeers
+    readSignal <- newMVar remoteHosts
```

Comment:
I do not understand the need for two variables here? Why can't one `TVar` suffices? Having to maintain consistency between two variables, and, outside of STM, seems error-prone.

Reply:
context:
- there is a thread `connect` in the background awaiting to receive a change on the MVar to continue
- there is another thread on `race_` awaiting to receive a change as well
- function `getPeers` on network expects to have a value in the MVar to read from

so, because of the mechanism used here to coordinate the reconnection and release of resources, getPeers will never yield a value but rather wait forever.

Reply:
`getPeers` is competing with `connect`, once `connect` takes the signal `getPeers` stops working.

Reply:
I think there's a bit of confusion here. That all those things are competing is *a good thing*, because it enforces a single-source of truth. 

Some things to note about `TVar` / `MVar`: 

- You can read the content of a variable without necessarily emptying it. So that means, a reader can almost always get access to what's stored without disturbing processes that may be waiting on that variable to change. 

- When working with STM, retrying is actually passive and cheap. For example, the following code:

   ```hs
   do
       atomically $ do 
           new <- readTVar var 
           check (new /= old) 
   ```
   
   will pretty much idle for as long as the content of `TVar` doesn't change. There's no active retry loop, rather, the runtime puts the thread at sleep and awaits for a change in the variable. 


This is seemingly how I would suggest to implement this:

- Store peers as `Set` (it makes comparison easier, and I don't think we rely on any particular ordering?) 
- Use only a single `TVar` (which may contains an empty set)
- Have the `connect` function starts by reading the content of the variable, and restart as soon as it changes to a different value. 

This way, there's no need to manage two sets of variables, and we can nicely leverage the runtime and behavior around software transactional memory to get something that is both elegant and efficient :blush: